src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1140,9 +1142,280 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
+
+            var callerSourceLocation = enableCallerInfo ? GetCallerLocation(syntax) : null;
+            BoundExpression defaultValue;
+            bool allowFailingConversion;
+            if (defaultConstantValue == ConstantValue.Unset)
+            {
+                // This is only expected to occur in recursive error scenarios, for example: `void F(object param = F()) { }`
+                defaultValue = new BoundDefaultExpression(syntax, parameterType) { WasCompilerGenerated = true };
+                allowFailingConversion = false;
+            }
+            else if (callerSourceLocation is object && parameter.IsCallerLineNumber)
+            {
+                int line = GetCallerLocation(syntax).SourceTree.GetDisplayLineNumber(callerSourceLocation.SourceSpan);
+                defaultValue = new BoundLiteral(syntax, ConstantValue.Create(line), Compilation.GetSpecialType(SpecialType.System_Int32)) { WasCompilerGenerated = true };
+                allowFailingConversion = true;
+            }
+            else if (callerSourceLocation is object && parameter.IsCallerFilePath)
+            {
+                string path = callerSourceLocation.SourceTree.GetDisplayPath(callerSourceLocation.SourceSpan, Compilation.Options.SourceReferenceResolver);
+                defaultValue = new BoundLiteral(syntax, ConstantValue.Create(path), Compilation.GetSpecialType(SpecialType.System_String)) { WasCompilerGenerated = true };
+                allowFailingConversion = true;
+            }
+            else if (callerSourceLocation is object && parameter.IsCallerMemberName)
+            {
+                var memberName = containingMember.GetMemberCallerName();
+                defaultValue = new BoundLiteral(syntax, ConstantValue.Create(memberName), Compilation.GetSpecialType(SpecialType.System_String)) { WasCompilerGenerated = true };
+                allowFailingConversion = true;
+            }
+            else if (defaultConstantValue == ConstantValue.NotAvailable)
+            {
+                // There is no constant value given for the parameter in source/metadata.
+                if (parameterType.IsDynamic() || parameterType.SpecialType == SpecialType.System_Object)
+                {
+                    // We have something like M([Optional] object x). We have special handling for such situations.
+                    defaultValue = GetDefaultParameterSpecialNoConversion(syntax, parameter, Compilation);
+                }
+                else
+                {
+                    // The argument to M([Optional] int x) becomes default(int)
+                    defaultValue = new BoundDefaultExpression(syntax, parameterType) { WasCompilerGenerated = true };
+                }
+                allowFailingConversion = false;
+            }
+            else if (defaultConstantValue.IsNull)
+            {
+                defaultValue = parameterType.IsValueType || (parameterType.IsNullableType() && parameterType.IsErrorType())
```

Comment:
>parameterType.IsNullableType() && parameterType.IsErrorType() [](start = 61, length = 61)

When is this true?

Reply:
It seems like this changes behavior for scenarios where the type argument to `System.Nullable<T>` is a missing (error) type. (the condition returns true, and we give a default literal instead of a null literal.)

I think I'm happier with changing this condition to say: if we know `null` is a valid literal of the parameter type, then use it, otherwise use `default`, since it ends up producing more "idiomatic" argument nodes in my view.

Reply:
I ended up going with using `BoundDefaultExpression` because it simplifies the code, it produces simpler operation trees, and you don't have to worry about wrapping the null literal nodes in conversions..